### PR TITLE
Fixed #17562 Ride exit next to park boundary exploit

### LINKED ISSUE #17562: Ride exit next to park boundary Exploit
### Operating System

Windows 11

### OpenRCT2 build

0.4.1-27

### Describe the issue

By placing a ride exit next to the park boundary you can trap guess outside of the park but they are still counted as guess inside the park. Also because the guess are not inside the park their happiness does not drop.

### Area(s) with issue?

The bug also exists in RCT2 (vanilla)

### Steps to reproduce

1. Create a new ride
2. Place exit on park boundary facing towards outside the park
3. Open ride, now it traps guess outside the park

### Attachments

bug

### PR DESCRIPTION
**Pull request by Kuhnovic**
amr4b5gRUX

Added a check to prevent building exits and entrances that are facing the park fence. Technically only required for exits, but added entrances for a more what is IMO a more consistent user experience. This fixes https://github.com/OpenRCT2/OpenRCT2/issues/17562

### CODE REVIEW

**Comment by ocalhoun6** [comment]
Just wanted to point out that there might be a specific edge case where this would be particularly annoying.

If the land outside the park boundary was for sale and you planned on building the ride first, then buying land for the entrance and exit paths. It's especially likely to happen in a scenario where you need to buy land a lot but have very little money to do so -- so you're only buying one tile at a time and _only_ the tiles you need.

Of course, in that case you can overcome it by just buying the extra tiles of land first, but it could be annoying to need to do so, especially since you'll have to close the construction window in order to do so.

File: src/openrct2/actions/RideEntranceExitPlaceAction.cpp [hunk 1]
```diff
@@ -97,6 +97,11 @@ GameActions::Result RideEntranceExitPlaceAction::Query() const
         return GameActions::Result(GameActions::Status::NotOwned, errorTitle, STR_LAND_NOT_OWNED_BY_PARK);
     }
 
+    if (!map_is_location_in_park(_loc - CoordsDirectionDelta[_direction]))
+    {
+        return GameActions::Result(GameActions::Status::NotOwned, errorTitle, STR_FACING_PARK_FENCE);
+    }
+
```
Comment (Broxzier) [inline src/openrct2/actions/RideEntranceExitPlaceAction.cpp:104]: Can't test right now. Does this account for tiles with partial ownership? If so, I think it'd make sense to check if the entrance/exit height is at least the minimum building height there. The problem then is that when there is no path, guests will fall outside the park range anyway.

**Comment by Broxzier** [comment]
Wouldn't it be better to just prevent guests from walking over the park edge, even if the tile they come from is occupied. Perhaps even make them walk through the ride exit in such cases.

File: src/openrct2/actions/RideEntranceExitPlaceAction.cpp (hunk 1 above)
Comment (Kuhnovic) [reply src/openrct2/actions/RideEntranceExitPlaceAction.cpp:104]: The map_is_location_owned function seems to take that into account:

image

Do you know of a park that has those specific building rights?

**Comment by Kuhnovic** [comment]
@ocalhoun6 The situation you describe might be annoying but probably a bit of an edge case (pun intended). Also, the first part of the error message is shared for all entrance / exit build actions, I only added a new string that appends the second park.

@Broxzier Allowing peeps to walk through the park edge / sides-of-ride-exit would look strange. Also, what happens if both tiles next to the exit are also blocked? I can also wonder how nice it would play with other code that might expect peeps to not enter blocked tiles / directions. IMO it's better to just prevent the whole thing from happening in the first place :)

**Comment by rik-smeets** [comment]
I tested this and fixed the implementation. Placing entrance/exit building is: 
- Not allowed when the land next to it is not owned and there's no partial ownership either;
- Not allowed when there can be no path built connecting it, even if there are building rights.

Everything can be tested quickly in Pokey Park for example. 

I think I could approve this PR but am not quite happy with the error message yet. 'Facing park edge!' doesn't really explain every possible scenario, since sometimes you _can_ build whilst facing the park edge (in case of partial ownership) which could be confusing. Maybe something like 'No path connection possible!'? What do you think @Broxzier, @Gymnasiast ?

**Comment by IntelOrca** [comment]
> Just wanted to point out that there might be a specific edge case where this would be particularly annoying.

Perhaps the error could be ignored if "disable clearance" is enabled.

**Comment by Broxzier** [comment]
> I think I could approve this PR but am not quite happy with the error message yet. 'Facing park edge!' doesn't really explain every possible scenario, since sometimes you can build whilst facing the park edge (in case of partial ownership)

The message seems alright to me. Park fences are placed on the tiles edge outside of the park, so in case the tile is at the same level, you wouldn't be able to immediate build a path anyway. Perhaps it could be reworded better though, to "Cannot build a path from this position".

In which case it might actually be better to check if something can be _constructed_ at the tile next to the entrance/exit.

https://github.com/OpenRCT2/OpenRCT2/blob/406b73d9048301810f915b57cff2f0e40146a79e/src/openrct2/actions/FootpathPlaceFromTrackAction.cpp#L145-L146

Then it'll already handle clearance cheat checks correctly too:

https://github.com/OpenRCT2/OpenRCT2/blob/406b73d9048301810f915b57cff2f0e40146a79e/src/openrct2/world/ConstructionClearance.cpp#L134-L138

**Comment by rik-smeets** [comment]
I can't get it to work properly with `MapCanConstructWithClearAt`, I just keep being able to build the entrance / exit buildings where they should not be allowed. Could you perhaps give it a go, @Broxzier (or someone else)?

**Comment by Broxzier** [comment]
Sure, I'll give it a go in the next few days.

**Comment by Kuhnovic** [comment]
Hey guys, thanks for picking up where I left off. I recently became a father so my priorities have shifted quite a bit hehe. Don't worry, my son will play OpenRCT before he can walk ;)

**Review by Broxzier — APPROVED** [review]

**Comment by Broxzier** [comment]
Congrats on becoming a father, @Kuhnovic!

Can someone give this PR a spin to test for side effects? This now touches a generic function that is used by other game actions too.

**Review by rik-smeets — CHANGES_REQUESTED** [review]
@Broxzier It doesn't work properly yet. I can build exits like this in Pokey Park:
image

While I own rights to build paths next to the entrance/exits, there's no way to actually build them at that specific location - you have to be higher up so you can build across the park fence.

**Comment by ZehMatt** [comment]
@Broxzier wanna take this one over?

**Comment by Broxzier** [comment]
Sure, I'll try to fix the issue that Rik found.

**Comment by 733737** [comment]
will this make it into 0.4.5?

**Comment by rik-smeets** [comment]
Probably not, it isn't really high priority. But let's keep this one open though, it'd be nice to get this one through somewhere in the coming months.

**Comment by BarryJRowe** [comment]
I don't think this is a good fix for the issue, as it may break existing custom scenarios that have tight space requirements for either fitting in enough long, high excitement rides, or those that need to place a few rides like to increase the soft guest cap. A better solution would be to just remove the guests that leave the park this way, or mark them as casualties so the park rating would be affected.

**Comment by ZehMatt** [comment]
> I don't think this is a good fix for the issue, as it may break existing custom scenarios that have tight space requirements for either fitting in enough long, high excitement rides, or those that need to place a few rides like to increase the soft guest cap. A better solution would be to just remove the guests that leave the park this way, or mark them as casualties so the park rating would be affected.

It should be impossible for peeps to leave the park because of an ride exit like that, this is a game breaking thing and deleting peeps is a really bad way to compensate for this sort of glitch which will also add technical debt due to having a new check that removes peeps which got outside of the park without leaving the park by normal means. If one absolutely must do this then there is the tile inspector or cheats. We also can't really account for every scenario ever made that requires to exploit some of the weaker designs of the game, there has to be a line drawn at some point.

**Comment by ocalhoun6** [comment]
It will still be possible for peeps to leave the park through a ride exit, simply by placing a ride exit over construction rights.

**Comment by ageekhere** [comment]
How about, 
Option1 - if the door of the exit is not facing a path that is within the park the ride cannot be opened.
Option2 - if the door of the exit is facing a park fence then the ride cannot be opened.

**Comment by ocalhoun6** [comment]
>Option1 - if the door of the exit is not facing a path that is within the park the ride cannot be opened.

This could impact people who are using rides as scenery? (Though they'd usually use test mode for that, not open, so it won't break down.) It will also impact people who want to use a ride for music only -- the music will only play in open mode, not test mode. It would force them to build a path at the exit before they could do these things. That's not a *huge* deal, but it is kind of an unnecessary annoyance.

It could cause some problems in edge cases where players *want* to drop peeps -- into the void, into water to drown them, onto open land because they *want* lost peeps for whatever reason, etc. That will become impossible.

It could be annoying sometimes if you're doing things in the 'wrong' order and want to open a ride before building the exit line. This could be something you'd actually want to do if it has a long ride duration and you want to get peeps started cycling through the ride before building the exit line in order to save a little time. Could be even more relevant to multi-station transport rides and such -- there have been occasions where I had a miniature railroad with 4 stations, but only 2 or 3 stations were actually connected to paths yet, because I hadn't expanded the rest of the path system to that part of the park yet. It would be annoying if the railroad refused to open until I had paths to all exits.

Overall,  Option 1 seems like a really big and invasive change to gameplay ... just to solve one very rare edge case of a ride exit dumping peeps outside of the park. Seems to cause more problems than it solves.

------

Option2 - if the door of the exit is facing a park fence then the ride cannot be opened.

With this option, you could still dump peeps outside the park by building the exit above construction rights.

But that's a difficult problem to solve, and probably outside of the scope of this anyway. Because you can also do this by building a path above construction rights, filing the path with peeps, and then deleting the path.

Overall, Option 2 seems to be the far more elegant and simple solution. Not a *perfect* solution, perhaps, but it would at least mostly fix the problem, and it wouldn't be too intrusive otherwise -- it will only affect what it's meant to affect, not a lot of other things like Option 1.

**Comment by ageekhere** [comment]
> Overall, Option 2 seems to be the far more elegant and simple solution

I agree that I like option 2 better, so in this example the blue merry go round would be able to open because the door is clear even though the side of the exit is touching the park fence. Where the yellow merry go round would not be able to open.

Screenshot 2023-06-06 080438

There are a few things that could also be tried here, 
1. The ride exit could rotate to face a valid square
exit

2. If there are no valid squares then the door is removed and the ride cannot be set to open, if opening a save file which had a ride already opened with an invalid ride exit then the ride would closed, a warning message be given to the player and the guest that are on the ride at the time could exit through the ride entrances as a one off, as far as the guests that are outside the park on loading a save I am not sure if they should be removed or be allowed to walk home.
  
Screenshot 2023-06-06 082331

If i missed any other edge case scenarios lets document them here.

I also thought of an Option 3, just count guest that are outside of the park but on the map for park rating.

**Comment by ageekhere** [comment]
This should not be closed yet.

**Comment by ageekhere** [comment]
Is there an update on this fix?

**Comment by fathertimejr** [comment]
I don't get this one.  Like... it's not solving a problem, just reducing player freedom.  No one is doing this on accident, and if they did, having to pick up each guest is an appropriate consequence (like forgetting an exit path).  Calling it an exploit calls to mind competitive play (speedrunning); if someone found it useful for that, the rules are: "If using OpenRCT2, you may use any official release [...]"--i.e. previous versions can be used.  There's positives and negatives to changing how the game works each way you might change it (like messing with BarryJRowe's custom scenarios); some ideas here are not functionally different from "extra difficulty options" ( https://github.com/OpenRCT2/OpenRCT2/pull/20170 )

I'm not saying the 1999 way is the best.  It irritated me that guests won't use the construction righted footpaths in Urban Park.
Also, I've wished butting a ride exit up to the park entrance would send guests on their way home rather than trapping them in the gate.  So... not necessarily against change, but the current solution is just an attack on how other people play the game.  If you don't like what happens when you put a ride exit against a park fence _don't do it_.  

I'm mad even though this change probably won't affect me.

**Comment by ocalhoun6** [comment]
And most speedrunners *already* use older versions anyway, because the older versions have slightly less overhead and run a little faster when fast-forwarded.
